src/filters/H5Zbzip2.jl

```diff
@@ -0,0 +1,203 @@
+        stream.bzalloc = C_NULL
+        stream.bzfree = C_NULL
+        # BZ2_bzDecompressInit
+        ret = CodecBzip2.decompress_init!(stream, 0, false)
+        if ret != CodecBzip2.BZ_OK
+            errror("H5Zbzip2: bzip2 decompress start failed with error $ret.")
+        end
+
+        stream.next_out = outbuf
+        stream.avail_out = outbuflen
+        stream.next_in = unsafe_load(buf)
+        stream.avail_in = nbytes
+
+        cont = true
+
+        while cont
+            # BZ2_bzDecompress
+            ret = CodecBzip2.decompress!(stream)
+            if ret < 0
+                error("H5Zbzip2: bzip2 decompression failed with error $ret.")
+            end
+            cont = ret != CodecBzip2.BZ_STREAM_END
+            if cont && stream.avail_out == 0
+                # Grow the output buffer
+                newbuflen = outbuflen * 2
+                newbuf = Libc.realloc(outbuf, newbuflen)
+                if newbuf == C_NULL
+                    error("H5Zbzip2: memory allocation failed for bzip2 decompression.")
+                end
+                stream.next_out = newbuf + outbuflen
+                stream.avail_out = outbuflen
+                outbuf = newbuf
+                outbuflen = newbuflen
+            end
+        end
+
+        outdatalen = stream.total_out_lo32
+        # BZ2_bzDecompressEnd
+        ret = CodecBzip2.decompress_end!(stream)
+        if ret != CodecBzip2.BZ_OK
+            error("H5Zbzip2: bzip2 compression end failed with error $ret.")
+        end
+    else
+        # Compress data
+
+        # Maybe not the same size as outdatalen
+        odatalen = Cuint(0)
+        blockSize100k = 9
+
+        # Get compression blocksize if present
+        if cd_nelmts > 0
+            blockSize100k = unsafe_load(cd_values)
+            if blockSize100k < 1 || blockSize100k > 9
+                error("H5Zbzip2: Invalid compression blocksize: $blockSize100k")
+            end
+        end
+
+        # Prepare the output buffer
+        outbuflen = nbytes + nbytes ÷ 100 + 600 # worse case (bzip2 docs)
+        outbuf = Libc.malloc(outbuflen)
+        @debug "Allocated" outbuflen outbuf 
+        if outbuf == C_NULL
+            error("H5Zbzip2: Memory allocation failed for bzip2 compression")
+        end
+
+        # Compress data
+        odatalen = outbuflen
+        r_odatalen = Ref{Cuint}(odatalen)
+        ret = BZ2_bzBuffToBuffCompress(outbuf, r_odatalen, unsafe_load(buf), nbytes,
+                                       blockSize100k, 0, 0)
+        outdatalen = r_odatalen[]
+        if ret != CodecBzip2.BZ_OK
+            error("H5Zbzip2: bzip2 compression failed with error $ret.")
+        end
+    end # if flags & API.H5Z_FLAG_REVERSE != 0
+    Libc.free(unsafe_load(buf))
+    unsafe_store!(buf, outbuf)
+    unsafe_store!(buf_size, outbuflen)
+
+    catch err
+        if outbuf != C_NULL
+            Libc.free(outbuf)
+        end
+        rethrow(err)
```

Comment:
I'm not sure how safe it is to throw inside a `@cfunction`.

Reply:
In general, it is probably better to `return -1`, since that would be a HDF5 error code.

Reply:
It seems to work when the HDF library is being called from Julia, but you are right.  It is possible this may be invoked outside of Julia.

Reply:
I'm generally not sure how safe it is (I think it corresponds to a longjmp in C), but since HDF5 has its own error handling machinery, we should generally use that

Reply:
I just checked the documentation and the correct error code is `0`:

> Returns the number of valid bytes of data contained in buf. In the case of failure, the return value is 0 (zero) and all pointer arguments are left unchanged.

https://docs.hdfgroup.org/hdf5/develop/_h5_zdevelop_8h.html#a130d8964a46667029c7d3c14572577c6

Reply:
Ah, you're right, i was looking at the wrong callback function:
https://portal.hdfgroup.org/display/HDF5/H5Z_REGISTER
In either case, I think it is generally safer to use HDF5 error handling machinery, as that will correctly clean up after itself.

Reply:
If you wanted nicer stacktraces in Julia, the correct way would probably be to push a custom error message onto the stack: https://portal.hdfgroup.org/display/HDF5/H5E_PUSH2.

Reply:
See https://discourse.julialang.org/t/what-happens-when-you-throw-an-error-from-a-cfunction/71510/2?u=simonbyrne.

Reply:
As of 6713689, the function does not emit an error. It is caught and `Csize_t(0)` is returned.